cmd/osmosisd/cmd/airdrop.go

```diff
@@ -0,0 +1,304 @@
+	cmd := &cobra.Command{
+		Use:   "export-airdrop-genesis [denom] [file] [totalAmount]",
+		Short: "Import balances from provided genesis to {FlagHome}/genesis.json",
+		Long: `Import balances from provided genesis to {FlagHome}/genesis.json
+Download:
+	https://raw.githubusercontent.com/cephalopodequipment/cosmoshub-3/master/genesis.json
+Init genesis file:
+	osmosisd init mynode
+Example:
+	osmosisd export-airdrop-genesis uatom ../genesis.json 100000000000000 --snapshot-output="../snapshot.json"
+	- Check genesis:
+		file is at ~/.osmosisd/config/genesis.json
+	- Snapshot
+		file is at "../snapshot.json"
+`,
+		Args: cobra.ExactArgs(3),
+		RunE: func(cmd *cobra.Command, args []string) error {
+			clientCtx := client.GetClientContextFromCmd(cmd)
+			depCdc := clientCtx.JSONMarshaler
+			cdc := depCdc.(codec.Marshaler)
+			aminoCodec := clientCtx.LegacyAmino.Amino
+
+			serverCtx := server.GetServerContextFromCmd(cmd)
+			config := serverCtx.Config
+
+			config.SetRoot(clientCtx.HomeDir)
+
+			denom := args[0]
+			filepath := args[1]
+			osdenom := "uosmo"
+			snapshotOutput, err := cmd.Flags().GetString(flagSnapshotOutput)
+			if err != nil {
+				return fmt.Errorf("failed to get snapshot directory: %w", err)
+			}
+
+			totalAmount, ok := sdk.NewIntFromString(args[2])
+			if !ok {
+				return fmt.Errorf("failed to parse totalAmount: %s", args[2])
+			}
+
+			genFile := config.GenesisFile()
+			appState, genDoc, err := genutiltypes.GenesisStateFromGenFile(genFile)
+			if err != nil {
+				return fmt.Errorf("failed to unmarshal genesis state: %w", err)
+			}
+
+			authGenState := authtypes.GetGenesisStateFromAppState(cdc, appState)
+
+			accs, err := authtypes.UnpackAccounts(authGenState.Accounts)
+			if err != nil {
+				return fmt.Errorf("failed to get accounts from any: %w", err)
+			}
+
+			jsonFile, err := os.Open(filepath)
+			if err != nil {
+				return err
+			}
+			defer jsonFile.Close()
+
+			byteValue, _ := ioutil.ReadAll(jsonFile)
+
+			var genStateV036 GenesisStateV036
+
+			setCosmosBech32Prefixes()
+			err = aminoCodec.UnmarshalJSON(byteValue, &genStateV036)
+			if err != nil {
+				return err
+			}
+
+			snapshot := []SnapshotFields{}
+			balanceIndexByAddress := make(map[string]int)
+			totalAtomBalance := sdk.NewInt(0)
+			for index, account := range genStateV036.AppState.Accounts {
+				totalAtomBalance = totalAtomBalance.Add(account.Coins.AmountOf(denom))
+				balanceIndexByAddress[account.Address.String()] = index
+				snapshot = append(snapshot, SnapshotFields{
+					AtomAddress: account.Address.String(),
+					AtomBalance: account.Coins.AmountOf(denom),
+					AtomPercent: sdk.NewDec(0),
+				})
+			}
+
+			for _, delegation := range genStateV036.AppState.Staking.Delegations {
+				address := delegation.DelegatorAddress
+				index, ok := balanceIndexByAddress[address.String()]
+				if !ok {
+					continue
+				}
+				sharesInt := delegation.Shares.RoundInt()
```

Comment:
Can't just use shares, as ratio between shares:Atoms is not always 1:1, because a validator could have gotten slashed.

Need to multiply by validator total shares and divide by a validator's total delegation shares.

See example from governance tally function here:
https://github.com/cosmos/cosmos-sdk/blob/master/x/gov/keeper/tally.go#L61